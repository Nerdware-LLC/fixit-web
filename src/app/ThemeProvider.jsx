import React from "react";
import { ThemeProvider as MTP, createTheme } from "@mui/material/styles";
import CssBaseline from "@mui/material/CssBaseline";
import { any } from "../types";

export const ThemeProvider = ({ children }) => {
  const theme = createTheme(THEME);

  return (
    <MTP theme={theme}>
      <CssBaseline />
      {children}
    </MTP>
  );
};

const THEME = {
  description: "dark",
  palette: {
    mode: "dark",
    primary: {
      light: "#f89a35",
      main: "#f78103", // Fixit orange
      dark: "#ac5a02"
    },
    secondary: {
      light: "#5381ff",
      main: "#2962ff", // blue
      dark: "#1c44b2"
    },
    success: {
      light: "#33d373",
      main: "#00c851",
      dark: "#008c38"
    },
    error: {
      main: "rgba(255, 115, 115, 0.87)"
    },
    info: {
      main: "#ff80ab" // AKA NOTIFICATION
    },
    text: {
      primary: "rgba(255, 255, 255, 0.87)",
      disabled: "rgba(255, 255, 255, 0.38)",
      hint: "rgba(255, 255, 255, 0.54)", // AKA PLACEHOLDER
      labels: "#f7a44a" // desaturated Fixit orange (CUSTOM KEY)
    },
    background: {
      paper: "#353535", // AKA SURFACE
      default: "#121212",
      backdrop: "rgba(0, 0, 0, 0.5)" // (CUSTOM KEY)
    }
  }
};

ThemeProvider.propTypes = {
  children: any.isRequired
};
