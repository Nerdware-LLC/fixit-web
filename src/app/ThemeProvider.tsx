import React from "react";
import MuiThemeProvider from "@mui/material/styles/ThemeProvider";
import CssBaseline from "@mui/material/CssBaseline";
import { createTheme, type PaletteOptions } from "@mui/material/styles";
import { themeStore } from ".";
import { any } from "../types";

export const ThemeProvider = ({ children }: { children: React.ReactNode }) => {
  const theme = themeStore.useSubToStore();

  return (
    <MuiThemeProvider theme={THEMES[theme]}>
      <CssBaseline />
      {children}
    </MuiThemeProvider>
  );
};

/**
 * `ThemeProvider` merges this theme object with the @mui/material
 * default theme:
 *
 * https://mui.com/material-ui/customization/default-theme/
 */
const THEMES = Object.fromEntries(
  ["DARK", "LIGHT"].map((themeModeName) => [
    themeModeName,
    createTheme({
      palette: {
        mode: themeModeName.toLowerCase() as PaletteOptions["mode"],
        primary: {
          light: "#f89a35",
          main: "#f78103", // Fixit orange
          dark: "#ac5a02"
        },
        secondary: {
          light: themeModeName === "DARK" ? "#00FFDC" : "#625AFA",
          main: themeModeName === "DARK" ? "#00FFDC" : "#625AFA",
          dark: themeModeName === "DARK" ? "#00FFDC" : "#625AFA"
        },
        success: {
          light: "#33d373",
          main: "#00c851",
          dark: "#008c38"
        },
        error: {
          main: "rgba(255, 115, 115, 0.87)"
        },
        info: {
          main: "#5796D6" // AKA NOTIFICATION
        },
        text: {
          primary: themeModeName === "DARK" ? "rgba(255, 255, 255, 0.87)" : "rgba(0, 0, 0, 0.87)",
          secondary: themeModeName === "DARK" ? "#00FFDC" : "#625AFA",
          disabled: "rgba(255, 255, 255, 0.38)",
          hint: "rgba(255, 255, 255, 0.54)", // AKA PLACEHOLDER
          labels: "#f7a44a" // desaturated Fixit orange (CUSTOM KEY)
        } as any, // <-- Custom theme properties are fine.
        background: {
          paper: themeModeName === "DARK" ? "#353535" : "#cccccc", // AKA SURFACE
          default: themeModeName === "DARK" ? "#121212" : "#f6f6f6",
          backdrop: "rgba(0, 0, 0, 0.5)" // (CUSTOM KEY)
        } as any // <-- Custom theme properties are fine.
      },
      typography: {
        fontFamily: '"Roboto", sans-serif'
      }
    })
  ])
);

ThemeProvider.propTypes = {
  children: any.isRequired
};
