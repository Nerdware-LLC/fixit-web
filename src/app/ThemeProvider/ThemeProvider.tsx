import CssBaseline from "@mui/material/CssBaseline";
import MuiThemeProvider from "@mui/material/styles/ThemeProvider";
import { usePageLayoutContext } from "@app/PageLayoutContext/usePageLayoutContext";
import { themeStore } from "@app/apolloCache/reactiveVars/themeStore";
import { THEMES } from "./themes";
import type { Theme } from "@mui/material/styles";

export const ThemeProvider = ({ children }: ThemeProviderProps) => {
  const themeName = themeStore.useSubToStore();
  const { isMobileUserAgent, isMobilePageLayout } = usePageLayoutContext();

  const theme: Theme = {
    ...THEMES[themeName || "DARK"], // A fallback themeName for added safety (will crash if not set)
    variables: {
      isMobileUserAgent,
      isMobilePageLayout,
    },
  };

  return (
    <MuiThemeProvider theme={theme}>
      <CssBaseline />
      {children}
    </MuiThemeProvider>
  );
};

export type ThemeProviderProps = { children: React.ReactNode };
