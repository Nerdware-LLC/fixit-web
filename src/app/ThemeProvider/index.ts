export { ThemeProvider } from "./ThemeProvider";
export * from "./themes";
