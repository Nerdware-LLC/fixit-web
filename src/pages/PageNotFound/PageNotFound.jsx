import React from "react";
import styled from "@emotion/styled";
import { useWebViewContext, Title, Text, Button } from "../../components";

export const PageNotFound = () => {
  const { webViewPostMessage } = useWebViewContext();

  const exitWebViewWithError = () => {
    webViewPostMessage({ error: "INVALID_ROUTE" });
  };

  return (
    <StyledPageNotFoundLayout>
      <Title>Whoops - this page doesn&apos;t seem to exist!</Title>
      <Text>
        <span role={"img"} aria-label={"x emoji"}>
          ‚ùå
        </span>{" "}
        404 - File or directory not found.
      </Text>
      <Text>
        The resource you are looking for might have been removed, had its name
        changed, or is temporarily unavailable.
      </Text>
      <Button
        label={"Take Me Back to Civilization!"}
        onClick={exitWebViewWithError}
        style={{ alignSelf: "center", fontWeight: "bold" }}
        variant={"contained"}
        color={"primary"}
        size={"large"}
      />
    </StyledPageNotFoundLayout>
  );
};

const StyledPageNotFoundLayout = styled.div`
  box-sizing: border-box;
  height: 100%;
  text-align: left;
  display: flex;
  flex-direction: column;
  padding: 30vh 15vw;
  align-items: center;
  justify-content: space-around;
`;
