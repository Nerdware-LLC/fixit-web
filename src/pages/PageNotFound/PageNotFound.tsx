import React from "react";
import { useNavigate } from "react-router-dom";
import { styled } from "@mui/material/styles";
import Button from "@mui/material/Button";
import Divider, { dividerClasses } from "@mui/material/Divider";
import Text, { typographyClasses } from "@mui/material/Typography";
import { DragonIcon } from "@components/Icons/DragonIcon";
import { Anchor } from "@components/Navigation/Anchor";
import { Link } from "@components/Navigation/Link";

export const PageNotFound = () => {
  const nav = useNavigate();

  return (
    <StyledDiv>
      <div className={classNames.headerContainer}>
        <Text variant="h2">
          <span className={classNames.text404}>404</span>Whoops - this page doesn&apos;t seem to
          exist!
        </Text>
      </div>
      <div className={classNames.bodyContainer}>
        <div>
          <Text>Looking for something in particular? Hopefully one of these links will help:</Text>
          <ul>
            {[
              { to: "/", label: "Home" },
              { to: "/pricing", label: "Pricing" },
              { to: "/privacy", label: "Privacy Policy" },
              { to: "/tos", label: "Terms of Service" },
            ].map(({ to, label }, index) => (
              <React.Fragment key={label}>
                {index !== 0 && <Divider orientation="vertical" flexItem />}
                <li key={to}>
                  <Link to={to}>{label}</Link>
                </li>
              </React.Fragment>
            ))}
          </ul>
        </div>
        <Text>
          Sure,{` `}
          <Anchor href="https://www.sparknotes.com/lit/fellowship/quotes/page/4/#:~:text=Quote%204,whether%20Strider%20is%20indeed%20Aragorn.">
            "Not all those who wander are lost"
          </Anchor>
          {`  `}
          and all that, but <i>you</i> - my friend - have wandered right off the map and straight
          into <em>"There be dragons" </em> territory! <DragonIcon />
        </Text>
      </div>
      <div className={classNames.buttonContainer}>
        <Button onClick={() => nav("/")} size="large">
          Take me home!
        </Button>
      </div>
    </StyledDiv>
  );
};

const classNames = {
  text404: "page-not-found-text-404",
  headerContainer: "page-not-found-header-container",
  bodyContainer: "page-not-found-body-container",
  buttonContainer: "page-not-found-button-container",
};

const StyledDiv = styled("div")(({ theme: { breakpoints, variables } }) => ({
  height: "100%",
  padding: "3rem 2rem",
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "flex-start",
  gap: "2rem",
  textAlign: "center",

  [breakpoints.up("md")]: {
    justifyContent: "center",
    padding: "4rem",
    gap: "4rem",
  },

  // ALL CHILD DIV-CONTAINERS (low-specificity)
  "& > div": {
    display: "flex",
    alignItems: "center",
    justifyContent: "center",
    maxWidth: "66rem",
  },

  // HEADER
  [`& > div.${classNames.headerContainer} > h2.${typographyClasses.root}`]: {
    ...(variables.isMobilePageLayout
      ? { fontSize: "2rem", lineHeight: "2.25rem" }
      : { fontSize: "3.25rem", lineHeight: "3.25rem" }),
    // "404" text
    [`& .${classNames.text404}`]: {
      opacity: 0.3,
      marginRight: "1rem",
    },
  },

  // BODY-CONTAINER
  [`& > div.${classNames.bodyContainer}`]: {
    flexDirection: "column",
    gap: "inherit",
    ...(variables.isMobilePageLayout
      ? { fontSize: "1rem", lineHeight: "1.5rem" }
      : { fontSize: "1.5rem", lineHeight: "2rem" }),

    [`& .${typographyClasses.root}`]: {
      fontSize: "inherit",
      lineHeight: "inherit",
    },

    "& > div > ul": {
      margin: "0.5rem 0 0 0",
      padding: 0,
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      flexDirection: variables.isMobilePageLayout ? "column" : "row",
      "& > li": {
        listStyleType: "none",
        margin: "0 1rem",
        "& > a": {
          textDecoration: "underline",
        },
      },
      [`& > .${dividerClasses.root}`]: {
        ...(variables.isMobilePageLayout && { display: "none" }),
      },
    },

    [`& > .${typographyClasses.root}`]: {
      whiteSpace: "pre-wrap",

      "& a,em": {
        whiteSpace: "nowrap",
        fontStyle: "italic",
      },
      "& svg": {
        ...(variables.isMobilePageLayout
          ? {
              height: "1rem",
              marginLeft: "0.15rem",
              transform: "translateY(2px)",
            }
          : {
              height: "1.5rem",
              marginLeft: "0.35rem",
              transform: "translateY(3px)",
            }),
      },
    },
  },

  // BUTTON
  [`& > div.${classNames.buttonContainer} > button`]: {
    marginTop: "0.5rem",
    fontStyle: "italic",
  },
}));
