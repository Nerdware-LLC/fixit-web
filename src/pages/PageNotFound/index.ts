export { Component } from "./PageNotFound";
// Exported as "Component" for react-router-dom lazy loading
