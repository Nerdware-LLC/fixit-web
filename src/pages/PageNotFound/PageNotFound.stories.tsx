import { withRootAppLayoutDecorator } from "@/../.storybook/decorators";
import { PageNotFound } from "./PageNotFound.jsx";
import type { Meta, StoryObj } from "@storybook/react";

const meta = {
  title: "Pages/PageNotFound",
  component: PageNotFound,
  decorators: [withRootAppLayoutDecorator],
  parameters: {
    layout: "fullscreen",
  },
} satisfies Meta<typeof PageNotFound>;

export default meta;

///////////////////////////////////////////////////////////
// STORIES

type Story = StoryObj<typeof meta>;

export const BasicDemo = {} satisfies Story;
