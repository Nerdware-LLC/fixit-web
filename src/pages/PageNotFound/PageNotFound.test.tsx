import renderer from "react-test-renderer";
import { MemoryRouter, Routes, Route } from "react-router-dom";
import { PageNotFound } from "./PageNotFound";
import { ThemeProvider } from "@app/ThemeProvider";

it("renders correctly", () => {
  expect(
    renderer
      .create(
        <ThemeProvider>
          <MemoryRouter initialEntries={["/test-page-not-found"]}>
            <Routes>
              <Route path="*" element={<PageNotFound />} />
            </Routes>
          </MemoryRouter>
        </ThemeProvider>
      )
      .toJSON()
  ).toMatchSnapshot();
});
