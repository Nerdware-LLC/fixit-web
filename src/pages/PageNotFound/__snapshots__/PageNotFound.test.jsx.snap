// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="css-15qcvv9"
>
  <p
    className="css-mafmc5"
  >
    Whoops - this page doesn't seem to exist!
  </p>
  <p
    className="css-v8rbnv"
  >
    <span
      aria-label="x emoji"
      role="img"
    >
      ‚ùå
    </span>

    404 - File or directory not found.
  </p>
  <p
    className="css-v8rbnv"
  >
    The resource you are looking for might have been removed, had its name changed, or is temporarily unavailable.
  </p>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeLarge MuiButton-containedSizeLarge css-1mo5pof-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "alignSelf": "center",
        "fontWeight": "bold",
      }
    }
    tabIndex={0}
    type="button"
  >
    Take Me Back to Civilization!
  </button>
</div>
`;
